lib/elixir/lib/calendar.ex

```diff
@@ -413,13 +523,79 @@ defmodule Date do
   """
   @spec compare(Calendar.date, Calendar.date) :: :lt | :eq | :gt
   def compare(date1, date2) do
-    case {to_erl(date1), to_erl(date2)} do
-      {first, second} when first > second -> :gt
-      {first, second} when first < second -> :lt
-      _ -> :eq
+    if Calendar.compatible_calendars?(date1.calendar, date2.calendar) do
+      case {to_rata_die(date1), to_rata_die(date2)} do
+        {first, second} when first > second -> :gt
+        {first, second} when first < second -> :lt
+        _ -> :eq
+      end
+    else
+      raise ArgumentError, """
+      cannot compare #{inspect date1} with #{inspect date2}:
+      This comparison would be ambiguous as their calendars have incompatible day rollover moments.
+      Specify an exact time of day (using `DateTime`s) to resolve this ambiguity.
+      """
+    end
+  end
+
+  def convert(%{calendar: calendar, year: year, month: month, day: day} = date, calendar), do: {:ok, date}
+  def convert(%{calendar: _, year: _, month: _, day: _} = date, target_calendar) do
```

Comment:
We need to decide if we want to support conversion of super types. For example, can I give a DateTime and pass it to `Date.convert/2`? I don't think we should because there will be loss of precision. This means we should match on the struct `%Date{}` instead of field names and make sure the spec expects `Date.t` instead of `Calendar.date`.